openshift/machine-config-operator
# OCPBUGS-33913, OCPBUGS-34261: CurrentImagePullSecret should be consumed by the MCD

**- What I did**

The CustomImagePullSecret that is referred to by a MachineOSConfig was not being used. This means that pulling an OS image from a private image registry will not work. This work does the following:

1. Migrates functions for changing the formatting of pull secrets from the BuildController package into controller/common, allowing reuse.
2. Adds secret volume mounts to the MachineConfigDaemon DaemonSet as well as template rendering logic to collect that information whenever a MachineOSConfig is changed.
3. Adds an MCD-side merge process that reads the secrets from the volume mounts, reads the secrets provided by the ControllerConfig, and then merges them into /etc/mco/internal-image-pull-secret.json.
4. Adds unit and e2e tests to validate that a secret referenced by a MachineOSConfig is written to the nodes' filesystem, overriding any values from the ControllerConfig.

**- How to verify it**

1. Bring up a cluster with the TechPreviewNoUpgrade feature-gate added.
2. Set up a MachineOSConfig to use a private image registry for both pushing and pulling.
3. Create a MachineConfigPool and a MachineOSConfig and wait for the build to complete. While this is occurring the MCD DaemonSet will restart since it will now have a secret mounted into it. If multiple MachineOSConfigs are present, multiple secret volumes will be mounted into the MCD pod, even if they are the same secret.
4. Move a node into the new MachineConfigPool and wait for it to drain and reboot.
5. When the node comes back up, it should be booted into the OS image that was pulled from the private image registry.

**- Description for the changelog**
CurrentImagePullSecret should be used by MachineConfigDaemon
